File: app/src/main/java/aipova/githubsearch/MainActivity.java
package aipova.githubsearch;

import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import aipova.githubsearch.dao.RepoDao;
import aipova.githubsearch.dao.RepoModel;

public class MainActivity extends AppCompatActivity {

    private RecyclerView reposRecyclerView;
    private ReposAdapter reposAdapter;
    private EditText editText;
    private RepoDao repoDao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        editText = (EditText) findViewById(R.id.search_text);

        reposRecyclerView = (RecyclerView) findViewById(R.id.repo_recycler_view);
        repoDao = new RepoDao();
        reposAdapter = new ReposAdapter(new ArrayList<RepoModel>());
        final LinearLayoutManager layoutManager = new LinearLayoutManager(getApplicationContext());
        reposRecyclerView.setLayoutManager(layoutManager);
        reposRecyclerView.setItemAnimator(new DefaultItemAnimator());
        reposRecyclerView.setAdapter(reposAdapter);
        reposRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            private boolean loading;
            private int pageCount = 1;
            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);

                int lastvisibleitemposition = layoutManager.findLastVisibleItemPosition();

                if (lastvisibleitemposition == reposAdapter.getItemCount() - 1) {

                    if (!loading) {
                        loading = true;
                        fetchData((++pageCount));
                    }
                }
            }

            private void fetchData(int page) {
                String searchString = editText.getText().toString();
                if (!searchString.isEmpty()) {
                    GetReposTask getReposTask = new GetReposTask();
                    getReposTask.execute(searchString, String.valueOf(page));
                }
                loading=false;
            }
        });
    }

    public void searchForRepo(View view) {

        reposAdapter.clear();
        try {
            InputMethodManager imm = (InputMethodManager)getSystemService(INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
        } catch (Exception e) {
            // TODO: handle exception
        }
        String searchString = editText.getText().toString();
        if (!searchString.isEmpty()) {
            GetReposTask getReposTask = new GetReposTask();
            getReposTask.execute(searchString, "1");
        }
    }

    class GetReposTask extends AsyncTask<String, Void, List<RepoModel>> {
        @Override
        protected List<RepoModel> doInBackground(String... params) {
            try {
                return repoDao.getRepoModelsFromApi(params[0], Integer.valueOf(params[1]));
            } catch (IOException apiAxeption) {
                Log.d(MainActivity.class.getName(), "Exception during api call: " + apiAxeption.getMessage());
            }
            return null;
        }

        @Override
        protected void onPostExecute(List<RepoModel> repoModels) {
            super.onPostExecute(repoModels);
            if (repoModels != null) {
                reposAdapter.addRepoModelsToList(repoModels);
                reposAdapter.notifyDataSetChanged();
            }
        }
    }
}

File: app/src/main/java/aipova/githubsearch/MainActivity.kt
package aipova.githubsearch

import aipova.githubsearch.data.Repo
import aipova.githubsearch.data.RepoDao
import android.content.Context
import android.os.AsyncTask
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.inputmethod.InputMethodManager
import kotlinx.android.synthetic.main.activity_main.*
import java.io.IOException

class MainActivity : AppCompatActivity() {

    private var reposAdapter = ReposAdapter(mutableListOf())
    private var repoDao: RepoDao = Injection.repoDao

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        searchBtn.setOnClickListener { searchForRepos() }
        setupRecyclerView()
    }

    private fun setupRecyclerView() {
        with(repoRecyclerView) {
            layoutManager = LinearLayoutManager(applicationContext)
            itemAnimator = DefaultItemAnimator()
            adapter = reposAdapter
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                private var loading: Boolean = false
                private var pageCount = 1
                override fun onScrollStateChanged(recyclerView: RecyclerView?, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    val lastVisibleItemPosition = (layoutManager as LinearLayoutManager).findLastVisibleItemPosition()
                    if (!loading && lastVisibleItemPosition == reposAdapter.itemCount - 1) {
                        loading = true
                        fetchData(++pageCount)
                        loading = false

                    }
                }
            })
        }

    }

    private fun fetchData(page: Int = 1) {
        val searchString = searchTxt.text.toString()
        if (!searchString.isEmpty()) {
            val getReposTask = GetReposTask()
            getReposTask.execute(searchString, page.toString())
        }
    }

    private fun searchForRepos() {
        reposAdapter.clear()
        hideKeyboard()
        fetchData()
    }

    private fun hideKeyboard() {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
    }

    inner class GetReposTask : AsyncTask<String, Void, List<Repo>>() {
        override fun doInBackground(vararg params: String): List<Repo>? {
            try {
                return repoDao.getRepoModelsFromApi(params[0], Integer.valueOf(params[1]))
            } catch (apiException: IOException) {
                Log.d(
                    MainActivity::class.java.name,
                    "Exception during api call: " + apiException.message
                )
            }

            return null
        }

        override fun onPostExecute(repoModels: List<Repo>?) {
            super.onPostExecute(repoModels)
            if (repoModels != null) {
                reposAdapter.addRepoModelsToList(repoModels)
                reposAdapter.notifyDataSetChanged()
            }
        }
    }
}
